Immutable TreeMap/TreeSet performance (SI-5331)

erikrozendaal opened the pull request:
As discussed on the scala-internals mailing list. There are some further improvements related to performance and code.

Note: This breaks test/files/run/t2873.scala, since it uses RedBlack#Empty for testing for the absence of a compiler bug. I don't know how to fix this test so that it still tests for the same compiler bug.

erikrozendaal committed:
RedBlack.scala: Change count from 'def' to 'val' in NonEmpty tree
to ensure TreeSet/TreeMap 'range' operations are O(log n) instead
of O(n).

erikrozendaal committed:
Use RedBlack.iterator to create iterators for TreeSet/TreeMap.
This turns iterator creation from an O(n) operation into an
O(log n) operation. Unfortunately, it halves actual iteration
speed (consuming the iterator fully), probably due to the many
by-name closures that are needed.

erikrozendaal committed:
Use custom implementation for iterating over RedBlack trees. Raw
performance is much better than '++' based iterator.

erikrozendaal committed:
Optimized implementations of head/headOption/last/lastOption for
TreeMap/TreeSet.

erikrozendaal committed:
Optimized implementation of init/tail for TreeSet/TreeMap.

erikrozendaal committed:
Implemented drop/take/slice/splitAt/dropRight/takeRight for
TreeMap/TreeSet by splitting the underlying RedBlack tree. This
makes the operation O(log n) instead of O(n) and allows more
structural sharing.

erikrozendaal committed:
Implemented takeWhile/dropWhile/span to use tree splitting. This
changes the operation from O(n log n) to O(n) and allows for more
structural sharing.

erikrozendaal committed:
Switched from isSmaller to ordering.

erikrozendaal committed:
Moved from implicit ordering value to implicit parameter.

erikrozendaal committed:
Moved from Empty case object to case class in preparation of moving
type parameter A.

erikrozendaal committed:
Moved type parameter A from RedBlack to Tree.

erikrozendaal committed:
Changed abstract class RedBlack to singleton object.

erikrozendaal committed:
Use single shared Empty instance across all RedBlack trees.

erikrozendaal committed:
Make sure the redblack test compiles and runs.

erikrozendaal committed:
Made RedBlack private to the scala.collection.immutable package.
Use ArrayStack instead of Stack in TreeIterator for slightly increased
performance.

erikrozendaal committed:
TreeMap/TreeSet no longer keep track of the size (the RedBlack tree
already does so).

erikrozendaal committed:
Performance improvements for iteration (foreach and iterator).

erikrozendaal committed:
Improved performance of RedBlack.NonEmpty.nth (helps take/drop/split/etc).

erikrozendaal committed:
Added some tests for TreeMap/TreeSet.

erikrozendaal committed:
Minimize number of calls to ordering.

paulp commented:
That's awesome.  Have you signed the SCA? http://www.scala-lang.org/sites/default/files/contributor_agreement.pdf I can't merge this much work until that's in place.

erikrozendaal commented:
Thanks for reminding me. Just mailed the signed SCA to Danielle at EPFL according to the instructions.

dcsobral commented on src/library/scala/collection/immutable/RedBlack.scala:
```diff
@@ -11,51 +11,54 @@
+    def update[B1 >: B](k: A, v: B1)(implicit ordering: Ordering[A]): Tree[A, B1] = blacken(upd(k, v))
+    def delete(k: A)(implicit ordering: Ordering[A]): Tree[A, B] = blacken(del(k))
+    def range(from: Option[A], until: Option[A])(implicit ordering: Ordering[A]): Tree[A, B] = blacken(rng(from, until))
+    def foreach[U](f: ((A, B)) =>  U)
```
This is wrong. By passing the Ordering on each method, you make it possible for the tree to be inconsistent. Sure, someone has to work at it, and having the class private helps prevent such misuse, but the very possibility indicates the ordering is in the wrong place. I really dislike this design.

Instead, the tree itself should have an Ordering (and getting A back), which kind of gets back to RedBlack being a class and its subclasses having a hidden pointer to it -- which would be required to access the Ordering. However, as long as TreeMap and TreeSet don't _extend_ RedBlack, it shouldn't be a problem. RedBlack becomes just a store for the Ordering shared by the nodes.

viktorklang replied:
I agree w Daniel

erikrozendaal replied:
This was another reason I feel RedBlack (as it currently is in this pull request) should be private[immutable]. This way TreeMap/TreeSet can ensure a consistent Ordering[A] is always used. 

Making RedBlack a class again could help with this. The costs are additional instances of this class (including the Empty nested object) and the additional $$outer pointer in the Tree nodes (this may or may not cause additional overhead, depending on alignment, etc). At least on MacOS X JDK 1.6.0_29 with compressed oops there was additional memory usage per node.

So definitely a tradeoff between performance and potential correctness.

viktorklang replied:
That's an understandable tradeoff. conserving memory is important for datastructures.

dcsobral commented on src/library/scala/collection/immutable/RedBlack.scala:
```diff
@@ -139,38 +143,40 @@ abstract class RedBlack[A] extends Serializable {
-    def iterator: Iterator[(A, B)] =
-      left.iterator ++ Iterator.single(Pair(key, value)) ++ right.iterator
+    override def foreach[U](f: ((A, B)) => U) {
+      if (left ne Empty.Instance) left foreach f
```
I'm pretty sure this is a pessimization. You introduced two test conditions that will be executed on every node except empty ones . Much better to have Empty do nothing.

erikrozendaal replied:
Strangely enough this makes a huge different in foreach. Looping over a random TreeMap[Int, Int] size 10,000:
- without guard conditions: ~3150 times per second
- with guard conditions: ~7100 times per second

For a TreeSet[Int] of the same size:
- without guard conditions: ~3000 times per second
- with guard conditions: ~6000 times per second

(TreeSet is slower than TreeMap, I suspect due to unboxing of Ints when invoking the supplied foreach body)

See https://gist.github.com/1507127 for the basic benchmark used (there must be an easier way to do this though).

This is why I feel using plain nulls and less polymorpishm will give another speed boost. For example, the TreeSet "head" method is still about 3 times slower than the java.util.Map "first", probably due to the need to a needed method call in RedBlack.NonEmpty.smallest.

viktorklang replied:
nullchecks are gigacheap

dcsobral replied:
That's bizarre. Are you sure this was the only thing changed on the benchmark? On the other hand, if are keeping the method this way, you might as well move the whole foreach to Tree, put a self condition for the empty tree, and make it final. Experiments with Range have shown this (making the method final in the class it is called) gives the best performance possible.

erikrozendaal replied:
I think pointer comparisons are just a lot faster than polymorphic calls. For a tree of size n there will be n+1 pointers to the empty node, so we save on half the calls to foreach. It may also help that the call sites to left.foreach() and right.foreach() are now monomorphic, maximizing the efficiency of the inline method call cache.

Moving the code to Tree and making it final is a bit harder though. Tree doesn't have access to key/value/left/right, so would need to do downcasting. Again, using null to represent the empty node and moving key/value/left/right to Tree would be optimal but is quite a complicated undertaking...

erikrozendaal committed:
Moved key/value/left/right fields up to NonEmpty class. Don't rely
on pattern matching for updating the tree.

erikrozendaal committed:
Implemented deletes without pattern matching.

erikrozendaal committed:
Implemented range without using pattern matching.

erikrozendaal committed:
Use null to represent empty trees. Removed Empty/NonEmpty classes.

erikrozendaal committed:
Optimize foreach and iterators.

erikrozendaal committed:
Move nth method to RedBlack. Inline factories for tree nodes.

erikrozendaal committed:
Add implementation notes. Consistently use eq/ne to compare with null.

erikrozendaal commented:
I've updated the RedBlack implementation to use `null` for representing the empty tree. Also the iterators were further optimized. You can find the implementation at https://github.com/erikrozendaal/scala/blob/redblack-with-null/src/library/scala/collection/immutable/RedBlack.scala.

Below you can see a performance comparison between the tree from this pull request (up to commit https://github.com/erikrozendaal/scala/commit/c51bdeaa6b85e132f24480fa93ded440c3511ab3), the null based redblack tree without inlining the accessors to the Tree fields, null based with inlining, `java.util.TreeMap` and `scala.collection.immutable.HashMap`. All measurements are based on a map with 10,000 random integer keys/value pairs. Run with Java 1.6.0_29 with -d64 -server. Numbers are operations per second.

<table>
<thead>
<tr><th></th><th>pull/82</th><th>with nulls, use getters</th><th>with nulls, inlined getters</th><th>java.util.TreeMap</th><th>immutable.HashMap</th></tr>
</thead>
<tbody>
<tr><th>lookup</th><td align="right">15,331,921</td><td align="right">25,107,208</td><td align="right">25,288,542</td><td align="right">24,927,772</td><td align="right">46,726,788</td></tr>
<tr><th>add</th><td align="right">1,876,683</td><td align="right">2,768,516</td><td align="right">5,216,653</td><td align="right"></td><td align="right">6,415,887</td></tr>
<tr><th>add & remove</th><td align="right">630,904</td><td align="right">1,261,497</td><td align="right">2,225,205</td><td align="right">3,193,698</td><td align="right">2,933,961</td></tr>
<tr><th>head</th><td align="right">30,097,999</td><td align="right">41,659,029</td><td align="right">69,986,842</td><td align="right">66,009,215</td><td align="right">46,456,097</td></tr>
<tr><th>last</th><td align="right">25,671,173</td><td align="right">36,402,421</td><td align="right">63,042,957</td><td align="right">67,400,887</td><td align="right"></td></tr>
<tr><th>iterator (create)</th><td align="right">3,850,423</td><td align="right">26,174,516</td><td align="right">27,073,711</td><td align="right">58,404,392</td><td align="right">180,170,441</td></tr>
<tr><th>iterator (first)</th><td align="right">3,741,934</td><td align="right">25,085,605</td><td align="right">25,207,521</td><td align="right">42,033,231</td><td align="right">46,981,884</td></tr>
<tr><th>iterator (all)</th><td align="right">3,594</td><td align="right">11,836</td><td align="right">11,925</td><td align="right">9,329</td><td align="right">4,652</td></tr>
<tr><th>foreach</th><td align="right">8,099</td><td align="right">15,375</td><td align="right">15,258</td><td align="right"></td><td align="right">7,058</td></tr>
<tr><th>drop(5)</th><td align="right">269,805</td><td align="right">669,568</td><td align="right">732,468</td><td align="right"></td><td align="right"></td></tr>
<tr><th>take(5)</th><td align="right">1,529,728</td><td align="right">3,544,943</td><td align="right">6,643,459</td><td align="right"></td><td align="right"></td></tr>
<tr><th>splitAt(size / 2)</th><td align="right">204,768</td><td align="right">451,053</td><td align="right">567,126</td><td align="right"></td><td align="right"></td></tr>
<tr><th>range(-1e8, 1e8)</th><td align="right">385,453</td><td align="right">712,859</td><td align="right">901,274</td><td align="right">53,214,983</td><td align="right"></td></tr>
<tr><th>range(-1e8, 1e8),headOption</th><td align="right">379,835</td><td align="right">700,441</td><td align="right">885,708</td><td align="right">12,531,721</td><td align="right"></td></tr>
</tbody>
</table>


As you can see the null based implementation is 2-4x faster than the one with the Empty singleton node. When inlining the accessors it is also very close to performance of the java version. The iteration optimization make it even faster than the Java TreeMap or the immutable.HashMap. The main area where the `java.util.TreeMap` still wins is with the range related operations. In the Java version `subMap` creates a view (O(1)) while the Scala version constructs a new tree (O(log n)).

Questions:
- Is this new code acceptable?
- If yes, shall I update this pull request or create a new one after this one is merged?

paulp commented:
It looks pretty good to me but I have to count on the others in this thread to help me spot issues as I have no time for the fine toothed comb treatment.  I'm sorry you had to resort to "@(inline @getter)" -- if you can put together some evidence tht direct field access is significantly faster, maybe we can convince martin to make it less difficult to achieve.

dcsobral commented:
On Thu, Jan 5, 2012 at 18:42, Erik Rozendaal
reply@reply.github.com
wrote:

> I've updated the RedBlack implementation to use `null` for representing the empty tree. Also the iterators were further optimized. You can find the implementation at https://github.com/erikrozendaal/scala/blob/redblack-with-null/src/library/scala/collection/immutable/RedBlack.scala.
> 
> Below you can see a performance comparison between the tree from this pull request (up to commit https://github.com/erikrozendaal/scala/commit/c51bdeaa6b85e132f24480fa93ded440c3511ab3), the null based redblack tree without inlining the accessors to the Tree fields, null based with inlining, `java.util.TreeMap` and `scala.collection.immutable.HashMap`. All measurements are based on a map with 10,000 random integer keys/value pairs. Run with Java 1.6.0_29 with -d64 -server. Numbers are operations per second.

I'm wondering if you shouldn't leave the current public implementation
of rb tree alone, and provide yours as a new private implementation,
plus the changes to TreeSet and TreeMap. The public implementation can
be deprecated, and we then look at the migration impacts of not having
it as ancestor to TreeSet/TreeMap.

## 

Daniel C. Sobral

I travel to the future all the time.

erikrozendaal commented:
> I'm wondering if you shouldn't leave the current public implementation
> of rb tree alone, and provide yours as a new private implementation,
> plus the changes to TreeSet and TreeMap. The public implementation can
> be deprecated, and we then look at the migration impacts of not having
> it as ancestor to TreeSet/TreeMap.

Yes, I thought about that too. Shouldn't be a problem and it may help avoid some level of source/binary compatibility. I wonder how many people really use RedBlack directly though. It has "implementation detail" written all over it (few docs, API not consistent with standard collections, etc). But maybe it's better to be safe than sorry :)

paulp commented:
I think that's a great idea.  Yes, we don't really have the luxury of assuming something isn't much used, we have to suffer the same restrictions on everything.  A new implementation makes my life much simpler.

ijuma commented:
I had a look at the benchmark and it seems to me that operations are being measured without using the result. Is this true or did I miss something? If that is true, then benchmark results are potentially inaccurate.

ijuma commented:
To give a concrete example:

```
bench("head", tree, count) { _.head }

def bench(name: String, tree: TreeSet[Int], count: Int)(block: TreeSet[Int] => Unit): Unit = {
  val s1 = time {
    for (_ <- 1 to count) {
      block(tree)
    }
  }
  println("%-25s: size: %7d: %7d iterations in %6.3f seconds (%12.2f per second)".format(name, tree.size, count, s1, count / s1))
}
```

erikrozendaal commented:
(Micro)benchmarking hotspot is indeed very hard, especially to get consistent results, even from run to run :(

I originally started with an implementation of bench that returned the last result. Later I switched to unit to reduce benchmark overhead. This didn't seem to change relative speeds between the various different benchmarks, so I'm pretty sure hotspot isn't totally optimizing away the code.

Using the following version of bench:

``` scala
def bench[A, B, C](name: String, tree: TreeMap[A, B], count: Int)(block: TreeMap[A, B] => C): C = {
  var result: C = null.asInstanceOf[C]
  val s1 = time {
    for (_ <- 1 to count) {
      result = block(tree)
    }
  }
  println("%-25s: size: %7d: %7d iterations in %6.3f seconds (%12.2f per second)".format(name, tree.size, count, s1, count / s1))
  result
}
```

Changes the `head` benchmarks to 31.5M / 44M for getters vs inlined getters respectively. So although the numbers change I don't think it affects the conclusion.

Finally, it is hard to tell the affect of micro-benchmark results on real world programs...

erikrozendaal committed:
Deprecate TreeMap.isSmaller and TreeSet.isSmaller.

These methods were used by the old RedBlack tree implementation,
but are no longer required and were not defined in any interface.
Use ordering or compare instead.

erikrozendaal committed:
Restore old RedBlack class to maintain backwards compatibility.

The class is marked as deprecated and no longer used by the TreeMap/TreeSet
implementation but is restored in case it was used by anyone else (since
it was not marked as private to the Scala collection library).

Renamed RedBlack.{Tree,RedTree,BlackTree} to Node, RedNode, and BlackNode
to work around name clash with RedBlack class.

erikrozendaal commented:
I restored the old implementation of RedBlack class in the commit https://github.com/erikrozendaal/scala/commit/f656142ddbcecfd3f8482e2b55067de3d0ebd3ce and marked the class as deprecated.

This required renaming the {Red,Black}Tree classes in the RedBlack object to avoid naming conflicts. I renamed them to {Red,Black}Node. Another option is to rename the RedBlack object to something else instead (so it is no longer a companion of the deprecated RedBlack class).

Let me know what you think and I'll adjust it and add it to this pull request.

erikrozendaal committed:
Renamed object RedBlack to RedBlackTree.

This more clearly separates the new implementation from the now
deprecated abstract class RedBlack and avoids naming conflicts
for the member classes.

erikrozendaal commented:
I added the null based implementation to this pull request. The old RedBlack class has also been restored and the new implementation is named RedBlackTree. I wonder if it is worth the effort to rewrite the history so that all binary compatible changes are at the start and the new RedBlackTree based implementation is added later. This would make integration with the next 2.9.x release easier. I could also just make a separate pull request for 2.9.x that only contains the binary compatible changes (mainly improvements to head/last/iterator/toStream).

ijuma commented:
The fact that micro-benchmarks on HotSpot are hard is more reason not to break the rules. There is enough uncertainty as it is without adding more. Also, if you want to reduce overhead, do not use Range (use a while loop). At least until the optimization efforts on that are finished.

erikrozendaal committed:
Tests for takeWhile/dropWhile/span.

Also simplified implementation of span to just use splitAt.

erikrozendaal commented:
Good point! I'll rerun the benchmarks using the following benchmark method:

``` scala

  def bench[A, B](name: String, tree: TreeMap[A, B], count: Int)(block: TreeMap[A, B] => Int => Int): Unit = {
    print("%-25s: ".format(name))
    val f = block(tree)
    var result = 0
    val elapsed = time {
      var i = 0
      while (i < count) {
        result += f(i)
        i += 1
      }
    }
    println("size: %7d: %7d iterations in %6.3f seconds (%12.2f per second): result = %s".format(tree.size, count, elapsed, count / elapsed, result))
  }

    val benches = Map[String, (TreeMap[java.lang.Integer, java.lang.Integer] => Int => Int)](
      "lookup" -> { tree => n => if (tree.contains(rnd(n))) 1 else 0 },
      "add" -> { tree => n => tree.updated(rnd(n), n).size },
      "add-remove" -> { tree => n => (tree.updated(rnd(n), n) - rnd(n)).size },
      "head" -> { tree => n => tree.head._1 },
      "last" -> { tree => n => tree.last._1 },
      "tail" -> { tree => n => tree.tail.size },
      "init" -> { tree => n => tree.init.size },
      "iterator" -> { tree => n => if (tree.iterator.hasNext) 1 else 0 },
      "iterator.next" -> { tree => n => tree.iterator.next._1 },
      "iterator.size" -> { tree => n => tree.iterator.size },
      // etc...
  )
```

I'll also run one benchmark at a time (in a single JVM), this seems to give best/most consistent results.

dcsobral commented:
You should use caliper to benchmark. Look at the fur foreach benchmark
project on my github as a basis. It's easy to compare multiple compiler
versions with it too.
Em 07/01/2012 16:50, "Erik Rozendaal" <
reply@reply.github.com>
escreveu:

> Good point! I'll rerun the benchmarks using the following benchmark method:
> 
> ``` scala
> 
>  def bench[A, B](name: String, tree: TreeMap[A, B], count: Int)(block:
> TreeMap[A, B] => Int => Int): Unit = {
>    print("%-25s: ".format(name))
>    val f = block(tree)
>    var result = 0
>    val elapsed = time {
>      var i = 0
>      while (i < count) {
>        result += f(i)
>        i += 1
>      }
>    }
>    println("size: %7d: %7d iterations in %6.3f seconds (%12.2f per
> second): result = %s".format(tree.size, count, elapsed, count / elapsed,
> result))
>  }
> 
>    val benches = Map[String, (TreeMap[java.lang.Integer,
> java.lang.Integer] => Int => Int)](
>      "lookup" -> { tree => n => if (tree.contains(rnd(n))) 1 else 0 },
>      "add" -> { tree => n => tree.updated(rnd(n), n).size },
>      "add-remove" -> { tree => n => (tree.updated(rnd(n), n) -
> rnd(n)).size },
>      "head" -> { tree => n => tree.head._1 },
>      "last" -> { tree => n => tree.last._1 },
>      "tail" -> { tree => n => tree.tail.size },
>      "init" -> { tree => n => tree.init.size },
>      "iterator" -> { tree => n => if (tree.iterator.hasNext) 1 else 0 },
>      "iterator.next" -> { tree => n => tree.iterator.next._1 },
>      "iterator.size" -> { tree => n => tree.iterator.size },
>      // etc...
>  )
> ```
> 
> I'll also run one benchmark at a time (in a single JVM), this seems to
> give best/most consistent results.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/82#issuecomment-3397375

erikrozendaal commented:
Here are the updated benchmark results. Complete results can be found at https://gist.github.com/1576263 (except for the java.util.TreeMap output, which accidentally overwrote). Numbers changed, sometimes radically, but I think the main conclusions are still the same.

<table><thead>
<tr><th></th><th>#82</th><th>with nulls, use getters</th><th>with nulls, inlined getters</th><th>java.util.TreeMap</th><th>immutable.HashMap</th></tr>
</thead><tbody>
<tr><th>lookup</th><td align="right">3,801,779</td><td align="right">7,362,195</td><td align="right">7,397,677</td><td align="right">7,572,085</td><td align="right">24,217,850</td></tr>
<tr><th>add</th><td align="right">1,192,201</td><td align="right">1,794,593</td><td align="right">2,427,549</td><td align="right"></td><td align="right">4,528,338</td></tr>
<tr><th>add & remove</th><td align="right">392,738</td><td align="right">867,402</td><td align="right">1,016,917</td><td align="right">3,528,667</td><td align="right">2,465,620</td></tr>
<tr><th>head</th><td align="right">28,917,103</td><td align="right">40,490,928</td><td align="right">61,891,427</td><td align="right">57,270,760</td><td align="right">41,516,223</td></tr>
<tr><th>last</th><td align="right">25,028,036</td><td align="right">32,569,793</td><td align="right">52,351,476</td><td align="right">56,938,347</td><td align="right"></td></tr>
<tr><th>iterator (create)</th><td align="right">4,667,903</td><td align="right">27,797,584</td><td align="right">27,981,034</td><td align="right">72,102,692</td><td align="right">366,207,914</td></tr>
<tr><th>iterator (consume first)</th><td align="right">4,567,211</td><td align="right">24,348,634</td><td align="right">24,282,580</td><td align="right">37,049,432</td><td align="right">42,535,377</td></tr>
<tr><th>foreach</th><td align="right">7,228</td><td align="right">12,541</td><td align="right">12,694</td><td align="right"></td><td align="right">9,231</td></tr>
<tr><th>iterator (consume all)</th><td align="right">3,583</td><td align="right">10,052</td><td align="right">9,895</td><td align="right">9,147</td><td align="right">6,202</td></tr>
<tr><th>drop(5)</th><td align="right">256,134</td><td align="right">632,225</td><td align="right">696,058</td><td align="right"></td><td align="right"></td></tr>
<tr><th>take(5)</th><td align="right">2,044,954</td><td align="right">4,463,086</td><td align="right">7,529,279</td><td align="right"></td><td align="right"></td></tr>
<tr><th>splitAt(size / 2)</th><td align="right">188,260</td><td align="right">367,508</td><td align="right">484,105</td><td align="right"></td><td align="right"></td></tr>
<tr><th>range(-1e8, 1e8).head</th><td align="right">382,497</td><td align="right">693,286</td><td align="right">880,981</td><td align="right">12,473,028</td><td align="right"></td></tr>
</tbody>
</table>

erikrozendaal committed:
Fix silly copy-paste error.

erikrozendaal committed:
Test for maximum height of red-black tree.

erikrozendaal commented:
Here are the results for OpenJDK build 1.7.0-u4-b05-20120106. It's nice to see the pattern match code got a decent speed boost (good news for more idiomatic Scala code with a lot of pattern matching). Also inlining of field access has less effect now. Still, the null based code is still the faster version of RedBlack.

PS Don't forget that the iterator based comparison is not fair between "#82" and the null based versions because additional algorithmic improvements were made in the meantime.
PPS All these numbers are running in 64-bit mode with compressed OOPS enabled, just like the previous 1.6.0_29 based results.

<table><thead>
<tr><th></th><th>#82 (pattern matching, singleton empty)</th><th>with nulls, use getters</th><th>with nulls, inlined getters</th><th>java.util.TreeMap</th><th>immutable.HashMap</th></tr>
</thead><tbody>
<tr><th>lookup</th><td>4,507,130</td><td>7,299,700</td><td>7,240,064</td><td>7,181,138</td><td>29,847,773</td></tr>
<tr><th>add</th><td>1,746,636</td><td>2,534,294</td><td>2,553,912</td><td></td><td>4,696,720</td></tr>
<tr><th>add & remove</th><td>756,158</td><td>1,072,570</td><td>1,060,522</td><td>3,496,455</td><td>2,608,751</td></tr>
<tr><th>head</th><td>25,772,815</td><td>43,935,453</td><td>58,751,122</td><td>59,795,623</td><td>43,097,982</td></tr>
<tr><th>last</th><td>22,547,783</td><td>37,729,653</td><td>52,286,158</td><td>59,816,398</td><td></td></tr>
<tr><th>iterator (create)</th><td>5,353,564</td><td>28,918,386</td><td>28,923,354</td><td>73,329,891</td><td>375,772,141</td></tr>
<tr><th>iterator (consume first)</th><td>5,572,964</td><td>23,698,292</td><td>23,899,985</td><td>35,495,586</td><td>42,454,982</td></tr>
<tr><th>foreach</th><td>8,054</td><td>12,523</td><td>12,411</td><td></td><td>9,541</td></tr>
<tr><th>iterator (consume all)</td><td>5,274</td><td>11,105</td><td>13,521</td><td>10,574</td><td>8,997</td></tr>
<tr><th>drop(5)</th><td>394,074</td><td>749,614</td><td>698,918</td><td></td><td></td></tr>
<tr><th>take(5)</th><td>3,680,581</td><td>7,920,021</td><td>7,778,135</td><td></td><td></td></tr>
<tr><th>splitAt(size / 2)</th><td>351,272</td><td>500,936</td><td>527,641</td><td></td><td></td></tr>
<tr><th>range(-1e8, 1e8).head</th><td>645,592</td><td>945,271</td><td>958,675</td><td>25,801,491</td><td></td></tr>
</tbody></table>

pavelpavlov commented:
Erik,

Please take a look at 'to' method implementation.
From the 4 subrange methods defined in scala.collection.generic.Sorted three (from, range, until) are implemented as single call to rangeImpl unlike 'to' which require two calls to 'rangeImpl' + creating and twice advancing subtree iterator in the worst case.

I wonder if it's worth to reimplement this method in TreeMap/TreeSet for better performance?

Also, it would be interesting to benchmark these four methods, especially "to(5)" vs. "until(6)".
I believe fast subrange creation is one of important scenarios for use of TreeSet/TreeMap.

erikrozendaal committed:
Optimized implementation of TreeMap/TreeSet#to method.

Performance of `to` and `until` is now the same.

erikrozendaal commented:
Hi Pavel,

I benchmarked the range operations and improved the implementation a bit such that `to` now has the same performance as `until`.

<table><thead>
<tr><th></th><th>old to</th><th>optimized to</th><th>java.util.TreeMap</th></tr>
</thead><tbody>
<tr><th>range(-1e8, 1e8).head</th><td>984,096</td><td>938,338</td><td>25,801,491</td></tr>
<tr><th>from(1).head</th><td>1,339,841</td><td>1,214,761</td><td>23,861,585</td></tr>
<tr><th>to(0).last</th><td>551,214</td><td>1,132,467</td><td>28,157,299</td></tr>
<tr><th>until(1).last</th><td>1,148,834</td><td>1,137,265</td><td>29,828,129</td></tr>
</tbody></table>


As you can see the `java.util.TreeMap` version is still much faster due to the use of views instead of building new trees. So to really get competitive with the Java version we need to support views (or ranged iterators, or something) and probably also implement methods like `lower`/`floor`/`ceiling`/`higher`, like `java.util.NavigableMap`.

pavelpavlov commented:
Hi Erik,

It's very cool that you could achieve such impressive performance improvements for your scenario.
However, I'm thinking about another performance-sensitive scenario: when we have zillion short-lived maps of moderate size each (say from 1 to 100 elements) instead of your scenario of having one map with zillion elements.

In that case another overheads come to the fore.
For example, looking into your last commit, you create 4 temporary objects on each call to `range`: two Options and two functions. Replacing option test here with null drops 2 of them and also gives faster calls to `after` and `before` in `rng`.

Also, you check `***Inclusive` invariants each time when `after` or `before` is called. It's better to create specialized closure for each case in `range`.

In general, it would be great to benchmark the scenario I mentionned above and get the real numbers.

Regards,
Pavel

erikrozendaal committed:
Custom coded version of range/from/to/until.

This avoids unnecessary allocation of Option and Function objects,
mostly helping performance of small trees.

erikrozendaal committed:
Custom implementations of drop/take/slice.

This mainly helps performance when comparing keys is expensive.

paulp commented:
FYI since everyone is doing such a fine job shaking the tree, I figure I'll let that continue a while longer.  (In other words, I haven't looked seriously at this patch and intend to let it mature as far as it can before I do.)

erikrozendaal commented:
Pavel,

I've some more improvements to `range`/`from`/`to`/`until` and `take`/`drop`/`slice`, basically by specializing `RedBlackTree.rng` for each of these cases. I'll try to get the benchmark numbers and patches posted here tonight CET.

erikrozendaal committed:
Removed TODOs.

erikrozendaal commented:
Here are the benchmark numbers related to the last few commits. I had to change the benchmarks somewhat to get decent numbers with small trees. So instead of getting the `head` or `last` element I just get the `size` of the resulting tree. For some benchmarks the random seed also had to be changed. All noted in the table below.

<table><thead>
<tr><th></th><th>master</th><th>default to (f26f610)</th><th>optimized to (00b5cb8)</th><th>specialized from/to/until/range (7824dbd)</th><th>specialized drop/take/slice (78374f3)</th></tr>
</thead><tbody>
<tr><th colspan="6">Size 10000, Seed 1234</th></tr>
<tr><th>range(-1e8, 1e8).head</th><td></td><td>984,096</td><td>938,338</td><td>1,274,907</td><td>1,274,907</td></tr>
<tr><th>from(1).head</th><td></td><td>1,339,841</td><td>1,214,761</td><td>1,358,472</td><td>1,358,472</td></tr>
<tr><th>to(0).last</th><td></td><td>551,214</td><td>1,132,467</td><td>1,208,318</td><td>1,208,318</td></tr>
<tr><th>until(1).last</th><td></td><td>1,148,834</td><td>1,137,265</td><td>1,246,551</td><td>1,246,551</td></tr>
<tr><th colspan="6">Size 10000, Seed 1234</th></tr>
<tr><th>drop(5).size</th><td></td><td>762,210</td><td>761,143</td><td>783,775</td><td>825,173</td></tr>
<tr><th>take(5).size</th><td></td><td>8,613,232</td><td>6,491,055</td><td>9,433,669</td><td>14,746,586</td></tr>
<tr><th>splitAt(size / 2).size</th><td></td><td>578,984</td><td>504,674</td><td>616,517</td><td>654,141</td></tr>
<tr><th>slice(3, size - 3).size</th><td></td><td>373,338</td><td>348,044</td><td>397,642</td><td>414,725</td></tr>
<tr><th colspan="6">Size 10, Seed 1234</th></tr>
<tr><th>range(-1e9, 1e9).size</th><td>8,213,661</td><td>11,389,161</td><td>8,291,250</td><td>14,930,028</td><td>15,113,249</td></tr>
<tr><th>from(1).size</th><td>8,630,657</td><td>13,060,927</td><td>11,276,415</td><td>14,540,512</td><td>14,371,185</td></tr>
<tr><th>to(0).size</th><td>1,052,989</td><td>3,840,600</td><td>11,524,856</td><td>14,487,387</td><td>14,262,957</td></tr>
<tr><th>until(1).size</th><td>7,775,261</td><td>11,946,964</td><td>11,365,586</td><td>14,217,502</td><td>14,295,698</td></tr>
<tr><th colspan="6">Size 10, Seed 1235</th></tr>
<tr><th>drop(5).size</th><td>281,202</td><td>11,900,665</td><td>10,404,720</td><td>12,763,176</td><td>14,397,612</td></tr>
<tr><th>take(5).size</th><td>347,416</td><td>11,831,718</td><td>10,036,927</td><td>15,386,880</td><td>16,705,482</td></tr>
<tr><th>splitAt(size / 2).size</th><td>587,840</td><td>5,683,290</td><td>4,796,227</td><td>7,146,472</td><td>8,209,405</td></tr>
<tr><th>slice(3, size - 3).size</th><td>320,606</td><td>6,529,863</td><td>5,025,384</td><td>9,531,225</td><td>11,963,432</td></tr>
</tbody></table>


I think I'm pretty much done with performance optimizations. I think it is now more interesting to discuss how we can get this patch into 2.10 and also see if there are binary compatible parts that can be put into 2.9. Maybe at least deprecate `RedBlack` starting with 2.9.2?

dcsobral commented:
A migration warning on TreeSet/TreeMap is necessary as well. I don't think there's any need to rush deprecation to 2.9.2 -- it can be deprecated on 2.10.

erikrozendaal commented:
So the migration warning should go into 2.9.x branch? Or 2.10? Warning about RedBlack inheritance, isSmaller, etc?

dcsobral commented:
Actually, I've changed my mind. I was thinking about warning about the fact that `TreeSet` no longer extends `RedBlack`, but any dependency on it will result in a compile error, so there's no need for it.

erikrozendaal commented:
So, anything else that needs to be done before merging this into 2.10?

paulp commented:
I will interpret the petering out of discussion as implied endorsement by the involved parties if someone doesn't say otherwise.  (It's still possible I personally will discover objections, but let's hope not.)
